mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java

```diff
@@ -144,21 +147,32 @@ public List<InputSplit> getSplits(JobContext context) {
     InputFormatConfig.InMemoryDataModel model =
         conf.getEnum(
             InputFormatConfig.IN_MEMORY_DATA_MODEL, InputFormatConfig.InMemoryDataModel.GENERIC);
-    try (CloseableIterable<CombinedScanTask> tasksIterable = scan.planTasks()) {
-      Table serializableTable = SerializableTable.copyOf(table);
-      tasksIterable.forEach(
-          task -> {
-            if (applyResidual
-                && (model == InputFormatConfig.InMemoryDataModel.HIVE
-                    || model == InputFormatConfig.InMemoryDataModel.PIG)) {
-              // TODO: We do not support residual evaluation for HIVE and PIG in memory data model
-              // yet
-              checkResiduals(task);
-            }
-            splits.add(new IcebergSplit(serializableTable, conf, task));
-          });
-    } catch (IOException e) {
-      throw new UncheckedIOException(String.format("Failed to close table scan: %s", scan), e);
+    final ExecutorService workerPool =
+        ThreadPools.newWorkerPool(
+            "iceberg-plan-worker-pool",
+            conf.getInt(
+                SystemConfigs.WORKER_THREAD_POOL_SIZE.propertyKey(),
+                ThreadPools.WORKER_THREAD_POOL_SIZE));
+    try {
+      scan = scan.planWith(workerPool);
```

Comment:
Hi, my test case fetch task to do the simple query, and `hive.fetch.task.conversion=more`

if `hive.fetch.task.conversion=none` and there will be no problem executing the mr task(I think the same goes for the tezTask)

Reply:
@lurnagao-dahua so you basically did the same as in https://github.com/apache/hive/commit/45867be6cb5308566e4cf16c7b4cf8081085b58c.

Reply:
> @lurnagao-dahua so you basically did the same as in [apache/hive@45867be](https://github.com/apache/hive/commit/45867be6cb5308566e4cf16c7b4cf8081085b58c).
Yes, applied to Iceberg.

Reply:
If this change is worked, i think it is because it close the thread pool after the planning like https://github.com/apache/hive/commit/45867be6cb5308566e4cf16c7b4cf8081085b58c, and this can ensure that each new fetch task can use the new thread pool with correct ugi.

Will do a test to double check this.

Reply:
Thank you for your reply!
I have conducted this test in my own internal environment and it is effective.

Reply:
````
set hive.fetch.task.conversion=more;
B user execution query select * from iceberg_tb_b;
A user execution query select * from iceberg_tb_a;
````
@lurnagao-dahua, could you please refactor code to be in sync with  https://github.com/apache/hive/commit/45867be6cb5308566e4cf16c7b4cf8081085b58c

Reply:
Hi,can be replicated and validated through this test case.
Thank you for your guidance. I will refactor the code.

Reply:
PS: i think it would be even better if we could push Hive-4.0 `IcebergInputFormat` and stuff into the Iceberg repo, but it was unrealistic so far.
cc @pvary

Reply:
> ```
> set hive.fetch.task.conversion=more;
> B user execution query select * from iceberg_tb_b;
> A user execution query select * from iceberg_tb_a;
> ```
> 
> @lurnagao-dahua, could you please refactor code to be in sync with [apache/hive@45867be](https://github.com/apache/hive/commit/45867be6cb5308566e4cf16c7b4cf8081085b58c)

Hi, Do you mean adjusting the code for iceberg, such as adding methods like planInputSplits?

Reply:
@deniskuzZ 
Hi,I initiated a merge request into my own project  [lurnagao-dahua_pull_2](https://github.com/lurnagao-dahua/iceberg/pull/2)
May I ask if you can check it for me.
Thank you very much!

Reply:
> ```
> set hive.fetch.task.conversion=more;
> B user execution query select * from iceberg_tb_b;
> A user execution query select * from iceberg_tb_a;
> ```
> 
> @lurnagao-dahua, could you please refactor code to be in sync with [apache/hive@45867be](https://github.com/apache/hive/commit/45867be6cb5308566e4cf16c7b4cf8081085b58c)

Hi,I initiated a merge request into my own project https://github.com/lurnagao-dahua/iceberg/pull/2
May I ask if you can check it for me.
Thank you very much!

Reply:
Looks good! After this change, each fetch task will new its own thread pool, so the ugi problem is gone.

But it would be good if you can try to add a unit test.

Reply:
> Looks good! After this change, each fetch task will new its own thread pool, so the ugi problem is gone.
> 
> But it would be good if you can try to add a unit test.

Hi, Thank you for your reply!
May I ask if you can give me some advice on writing unit test.

Reply:
> Looks good! After this change, each fetch task will new its own thread pool, so the ugi problem is gone.
> 
> But it would be good if you can try to add a unit test.

Hi,May I ask some questions about unit testing?
The Hadoop project has launched a MiniKdc service for testing ugi. Should I also do the same?
Looking forward to your reply very much!

Reply:
This test can also run in non-kerberos env, so i think you can use `org.apache.hadoop.security.UserGroupInformation` to login or get ugi info, then you don't need to launch the MiniKdc service.

Reply:
> This test can also run in non-kerberos env, so i think you can use `org.apache.hadoop.security.UserGroupInformation` to login or get ugi info, then you don't need to launch the MiniKdc service.


In the local file system, process users are used to read and write files, not ugi users. File permissions are determined by the permissions of the Linux system, so how to verify this issue?
Moreover, work_pool will shut down after execution and cannot obtain the actual ugi in the thread.
Can you provide further guidance on how to write this unit test case?
Looking forward to your reply very much!

Reply:
check `TestHiveCatalog#createTableAndVerifyOwner`, maybe that would be helpful. 
why do you need to verify `ugi`, you mentioned that without the fix test-case would fail

Reply:
how about something like
````
UserGroupInformation user1 = UserGroupInformation.getCurrentUser();
user1.doAs((PrivilegedAction<Key>)  () -> planInputSplits(...));
````

Reply:
Thank you for your reply!
the scope of the planInputSplits method is private.

Reply:
> check `TestHiveCatalog#createTableAndVerifyOwner`, maybe that would be helpful. why do you need to verify `ugi`, you mentioned that without the fix test-case would fail

Thank you for your reply!
In the local file system, process users are used to read and write files, not ugi users, so unit test will not failed.

Reply:
@deniskuzZ 
Hi,Could you please help me take a look at this test case?
Can I directly test the ugi of threads in the thread pool?
https://github.com/lurnagao-dahua/iceberg/pull/4/files
Looking forward to your reply very much!

Reply:
> This test can also run in non-kerberos env, so i think you can use `org.apache.hadoop.security.UserGroupInformation` to login or get ugi info, then you don't need to launch the MiniKdc service.

Thank you for your reply!
Sorry for bothering you.
May I ask if you could review my code again when you have some free time?
I would really appreciate it!